Release Management Notes (#5358)
plone/volto: docs/source/release-management-notes/index.md

---
myst:
  html_meta:

>     "description": "Release Process Information"

Reviewer: ```suggestion
    "description": "Release management notes for Volto"
```

>     "property=og:description": "Volto Release Notes for the Plone content management system"

Reviewer: ```suggestion
    "property=og:description": "Release management notes for Volto"
```

>     "property=og:title": "Volto Release Notes"

Reviewer: ```suggestion
    "property=og:title": "Volto release management notes"
```

    "keywords": "Volto, Plone, frontend, Release Notes, change log, changelog, change history"
---

> # Release Management Notes

Reviewer: ```suggestion
# Release management notes
```

> ## Definition of Breaking

Reviewer: ```suggestion
## Definition of breaking
```

### DOM structure and default styling

Any change done to the DOM structure of any Volto component that renders HTML is considered a breaking change.
This is because Volto projects rely on the DOM of the stable version they are using for styling and look and feel customizations.

The removal of any property or tag is also considered a breaking change.
This specially includes `className` props that would cause existing CSS to fail.

The addition of DOM is also considered a breaking change since it would potentially result in deliver unstyled or not customized in a Volto project.

```{note}
Exceptions: It is not considered a breaking change when the change to the DOM is the addition or introduction of a property. eg. the addition of a class name to a component or tag.
```

## Private API

The private API is everything that is considered "internal" to Volto and it is intended only to be used by Volto itself.

The following Volto components are considered Private API:

- Libraries used by Volto

### Example

If a bug is found in the text block and the solution involves to upgrade the Slate library, it won't be considered a breaking change.

> Even if that upgrade implies a breaking change on itself.

Reviewer: ```suggestion
This is true, even if that upgrade implies a breaking change in itself.

```

---
myst:
  html_meta:
    "description": "Release management notes for Volto"
    "property=og:description": "Release management notes for Volto"
    "property=og:title": "Volto release management notes"
    "keywords": "Volto, Plone, frontend, Release Notes, change log, changelog, change history"
---

# Release management notes

This chapter describes how maintainers manage the releases of Volto.

> 

Reviewer: ````suggestion

```{seealso}
Read the [Plone Release Schedule](https://plone.org/download/release-schedule) for details of how Volto's release management works in Plone.
```

````

This is because Volto projects rely on the DOM of the stable version for styling and look and feel customizations.

The removal of any property or tag is also considered a breaking change.
This especially includes `className` props that would cause existing CSS to fail.

The addition of DOM elements is also considered a breaking change, because it could result in delivering unstyled or uncustomized content in a Volto project.

```{note}
Exceptions: it is not considered a breaking change when the change to the DOM is the addition or introduction of a property, such as the addition of a class name to a component or tag.
```

(release-management-private-api)=

## Private API

The private API is everything that is considered "internal" to Volto and it is intended only to be used by Volto itself.

The following Volto components are considered Private API:

- CMS user interface fundamental components (`./src/components/manage`)
- CMS user interface control panels
- Libraries used by Volto

```{note}
In general, the CMSUI is not considered public API, because the customization/modification in a project is not usual, and the overall experience is tied to these components.

> So we prefer give preference to maintain a good user experience over not backporting changes that could be breaking to the product but could be beneficial for the vast majority of use cases.

Reviewer 2: "prefer give preference" is double.  I don't mind how to fix this, but this is one way:

```suggestion
So we give preference to maintain a good user experience over not backporting changes that could be breaking to the product but could be beneficial for the vast majority of use cases.
```

-   {term}`CMS` user interface fundamental components (`./src/components/manage`)
-   CMS user interface control panels
-   Libraries used by Volto

```{note}
In general, the CMS user interface is not considered public API, because the customization or modification in a project is not usual, and the overall experience is tied to these components.
Thus we give preference to maintaining a good user experience over not backporting changes that could be breaking to the product but could be beneficial for the vast majority of use cases.

For example, if a bug is fixed in the CMS user interface and it would be declared as breaking, it would never get backported to older versions.
Similarly for a useful improvement to any CMS user interface feature, or a new feature, they would never get backported to older versions.
```

## Examples of breaking and non-breaking changes

If there is a change in the CMS user interface that implies either a change of behavior or change of markup that improves the overall user experience, it's not a breaking change.
It will be properly identified in the {doc}`../release-notes/index` as a minor change.

> Any change to a control panel is a breaking change.

Reviewer: Addressed in https://github.com/plone/volto/pull/5358/commits/2fcdcc90fe1689574c2e60aada23c8fbd7d2f03e
